GVFS/GVFS.Common/Git/LibGit2Repo.cs

```diff
@@ -256,7 +228,7 @@ public static class Repo
                 [DllImport(Git2NativeLibName, EntryPoint = "git_repository_open")]
                 public static extern uint Open(out IntPtr repoHandle, string path);
 
-                [DllImport(Git2NativeLibName, EntryPoint = "git_tree_free")]
+                [DllImport(Git2NativeLibName, EntryPoint = "git_repository_free")]

```

Comment:
This is a very old bug. It wasn't actually freeing the repository handle, because the method only freed tree handles. Of course, we never ran this method unless we were killing the process, anyway.

However, it meant that the fix I was trying to do didn't actually work. (It does with this change.)

The way to test it is:

1. Have packfiles that will be expired during the packfile maintenance step.
2. Mount your repo, and hydrate a new file.
3. Wait for the libgit2 pool to dispose of all repos.
4. Run the packfile maintenance step.
5. Look at the `NumStaleIdxFiles` and `NumIdxDeletionsBlocked` values in the `PackfileMaintenanceStep_PerformMaintenance` log entry.

Before this change, `NumStaleIdxFiles` was zero and `NumIdxDeletionsBlocked` was growing. With this change, we could delete all stale files when the libgit2 repos are disposed.

If the packfile maintenance step runs soon after someone hydrated a file through ProjFS (different than a `git checkout` would fill a file), then we can still get blocked deletions. This could be an argument to make the disposal schedule more aggressive. We can quickly ramp up the pool as needed.

Reply:
Good find! Unfortunate that this went so long without being found - but I guess we only attempted to call it as we were killing the process in any case.

Besides your manual steps, I couldn't really think of a good way to add automated tests that we are calling this function correctly. I suppose you could open a repository, and then free it, and then check that handles to pack files were closed as expected, but I am not sure that is worth the effort.

Anyway - If you wanted to get this in sooner, we could target this fix against master by itself.

> This could be an argument to make the disposal schedule more aggressive.

Should the pack file maintenance step coordinate with the libgit2 repo pool? Maybe the pack file maintenance could indicate that it wants to run and have the mount process to clear out the libgit2 repo pool. It could still fail (if there is ongoing filesystem activity that requires libgit2 repositories), but it would remove 1 potential failure mode (existing libgit2 repos from previous activity).

Reply:
The packfile maintenance step has some things already in place to mitigate when there are open handles. Specifically, it cleans up stale IDX files on its own, including those from previous steps.

However, maybe that stale-IDX logic should be in the prefetch code, as it already has a lot of logic to clean up back packs.